qt/scientific_interfaces/Indirect/IndirectFittingModel.h

```diff
@@ -49,101 +57,111 @@ class MANTIDQT_INDIRECT_DLL IndirectFittingModel {
   virtual boost::optional<std::string> isInvalidFunction() const;
-  virtual std::size_t numberOfWorkspaces() const;
-  std::size_t getNumberOfSpectra(std::size_t index) const;
+  virtual DatasetIndex numberOfWorkspaces() const;
+  SpectrumRowIndex getNumberOfSpectra(DatasetIndex index) const;
+  SpectrumRowIndex getNumberOfDomains() const;
+  virtual SpectrumRowIndex getDomainIndex(DatasetIndex dataIndex,
+                                          WorkspaceIndex spectrum) const;
   std::vector<std::string> getFitParameterNames() const;
-  virtual Mantid::API::IFunction_sptr getFittingFunction() const;
+  virtual Mantid::API::MultiDomainFunction_sptr getFittingFunction() const;
 
   virtual std::vector<std::string> getSpectrumDependentAttributes() const = 0;
 
   void setFittingData(PrivateFittingData &&fittingData);
-  void setSpectra(const std::string &spectra, std::size_t dataIndex);
-  void setSpectra(Spectra &&spectra, std::size_t dataIndex);
-  void setSpectra(const Spectra &spectra, std::size_t dataIndex);
-  virtual void setStartX(double startX, std::size_t dataIndex,
-                         std::size_t spectrum);
-  virtual void setEndX(double endX, std::size_t dataIndex,
-                       std::size_t spectrum);
+  void setSpectra(const std::string &spectra, DatasetIndex dataIndex);
+  void setSpectra(SpectraNew &&spectra, DatasetIndex dataIndex);
+  void setSpectra(const SpectraNew &spectra, DatasetIndex dataIndex);
+  virtual void setStartX(double startX, DatasetIndex dataIndex,
+                         WorkspaceIndex spectrum);
+  virtual void setStartX(double startX, DatasetIndex dataIndex);
+  virtual void setEndX(double endX, DatasetIndex dataIndex,
+                       WorkspaceIndex spectrum);
+  virtual void setEndX(double endX, DatasetIndex dataIndex);
   virtual void setExcludeRegion(const std::string &exclude,
-                                std::size_t dataIndex, std::size_t spectrum);
+                                DatasetIndex dataIndex,
+                                WorkspaceIndex spectrum);
 
   virtual void addWorkspace(const std::string &workspaceName);
   void addWorkspace(const std::string &workspaceName,
                     const std::string &spectra);
-  void addWorkspace(const std::string &workspaceName, const Spectra &spectra);
+  void addWorkspace(const std::string &workspaceName,
+                    const SpectraNew &spectra);
   virtual void addWorkspace(Mantid::API::MatrixWorkspace_sptr workspace,
-                            const Spectra &spectra);
-  virtual void removeWorkspace(std::size_t index);
+                            const SpectraNew &spectra);
+  virtual void removeWorkspace(DatasetIndex index);
   virtual PrivateFittingData clearWorkspaces();
   void setFittingMode(FittingMode mode);
-  virtual void setFitFunction(Mantid::API::IFunction_sptr function);
+  virtual void setFitFunction(Mantid::API::MultiDomainFunction_sptr function);
```

Comment:
We can pass all of these in this file by ref, as we don't need to own a copy and when we pass it to the addFitFunction method it will automatically update it's ownership though taking a copy